numba/cuda/cudadrv/driver.py

```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+
+        if mapped:
+            self._attempt_allocation(allocator)
+        else:
+            allocator()
+
+        finalizer = _hostalloc_finalizer(self, pointer, size, mapped)
+        ctx = weakref.proxy(self.context)
+
+        if mapped:
+            mem = MappedMemory(ctx, pointer, size, finalizer=finalizer)
+            self.allocations[mem.handle.value] = mem
+            return mem.own()
+        else:
+            return PinnedMemory(ctx, pointer, size, finalizer=finalizer)
+
+    def mempin(self, owner, pointer, size, mapped=False):
+        """Implements the pinning of host memory.
+
+        It is recommended that this method is not overridden by EMM Plugin
+        implementations - instead, use the :class:`BaseCUDAMemoryManager`.
+        """
+        if isinstance(pointer, (int, long)):
+            pointer = c_void_p(pointer)
+
+        # possible flags are "portable" (between context)
+        # and "device-map" (map host memory to device thus no need
+        # for memory transfer).
+        flags = 0
+
+        if mapped:
+            flags |= enums.CU_MEMHOSTREGISTER_DEVICEMAP
+
+        def allocator():
+            driver.cuMemHostRegister(pointer, size, flags)
+
+        if mapped:
+            self._attempt_allocation(allocator)
+        else:
+            allocator()
+
+        finalizer = _pin_finalizer(self, pointer, mapped)
+        ctx = weakref.proxy(self.context)
+
+        if mapped:
+            mem = MappedMemory(ctx, pointer, size, owner=owner,
+                               finalizer=finalizer)
+            self.allocations[mem.handle.value] = mem
+            return mem.own()
+        else:
+            return PinnedMemory(ctx, pointer, size, owner=owner,
+                                finalizer=finalizer)
+
+    def reset(self):
+        """Clears up all host memory (mapped and/or pinned) in the current
+        context.
+
+        EMM Plugins that override this method must call ``super().reset()`` to
+        ensure that host allocations are also cleaned up."""
+        self.allocations.clear()
+        self.deallocations.clear()
+
+    @contextlib.contextmanager
+    def defer_cleanup(self):
+        """Disables cleanup of mapped or pinned host memory in the current
+        context.
+
+        EMM Plugins that override this method must obtain the context manager
+        from this method before yielding to ensure that cleanup of host
+        allocations is also deferred."""
+        with self.deallocations.disable():
+            yield
+
+
+class NumbaCUDAMemoryManager(HostOnlyCUDAMemoryManager):
+    """Internal on-device memory management for Numba. This is implemented using
+    the EMM Plugin interface, but is not part of the public API."""
+
+    def initialize(self):
+        # Set the memory capacity of *deallocations* as the memory manager
+        # becomes active for the first time
+        if self.deallocations.memory_capacity == _SizeNotSet:
```

Comment:
Now that `_SizeNotSet` is more int-like than before, I felt that `==` was more appropriate. This could get confused with a device that has a memory capacity of 0, but such a device doesn't exist. I'm happy to change it back to `is` if you'd prefer?

Reply:
I think your reasoning is fine. TBH, I'm more concerned about the module globals having a potential race condition on the type of `_SizeNotSet`!

Reply:
I'm not sure what you're referring to regarding a potential race condition?

Reply:
Oh that there's two definitions for the same global and its protected by virtue of the import lock impl. Not of concern for this PR.